### ISSUE TITLE
New --links parameter not working with mount?

### ISSUE THREAD

**Issue by ErAzOr2k** [rclone/rclone#2975]
#### What is the problem you are having with rclone?

I uploaded some symlinks with "rclone copy" and parameter "--links" to my gdrive remote.
As expected, files with the extension "rclonelink" are generated.

I now expect that the rclonelink files are not visible via rclone mount. Instead they should be translated to symlinks. But actually it is not the case. Only plain rclonelink files are visible.

Additionally I'm also not able to create symlink. Instead I get I/O error (see below).

#### What is your rclone version (output from `rclone version`)
rclone v1.46-DEV (latest commit as of today)

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Linux x64

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone --config /config/rclone.conf mount gdrive_decrypted:/ /mnt --allow-other --allow-non-empty --uid=1000 --gid=1000 --umask=007 --links --fast-list --rc--verbose

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
ln -s /tmp/testfile testlink
ln: testlink: I/O error

**Comment by ErAzOr2k** [comment]
Ok if I undetstand it correctly, it only works with local remote.
Is it possible to implement this feature for drive remote?

**Comment by ncw** [comment]
I think this would have to be a new flag for mount, say `--use-links` or something like that which inteprets the `.rclonelink` files and turns them back into symlinks.

Therefore you would be able to mount your drive remote, and have the symlinks appear as symlinks.

Is that what is required do you think?

**Comment by ncw** [comment]
This probably wouldn't be too tricky if someone wanted to have a go?

**Comment by ErAzOr2k** [comment]
> I think this would have to be a new flag for mount, say `--use-links` or something like that which inteprets the `.rclonelink` files and turns them back into symlinks.
> 
> Therefore you would be able to mount your drive remote, and have the symlinks appear as symlinks.
> 
> Is that what is required do you think?

yes, this is exactly the expected behavior :)

**Comment by felisucoibi** [comment]
I hope this i implemented soon, i would like too mount to follow symlinks instead of show .rclonesymlink and e able to reate symlinks in the cloud (internally .rlclonelink) but for mount a symbolik one.

**Comment by diamondsw** [comment]
I was just wondering about this myself. It would certainly be useful if the mount layer could seamlessly translate rclonelink files back into symlinks (and bonus points if creating symlinks on a rw rclone mount would create rclonelink files behind the scenes).

**Comment by thekiefs** [comment]
I would also love this feature!

**Comment by dantebarba** [comment]
any alternatives to this. Can we use mergerfs instead?

**Comment by kocane** [comment]
@ncw Any plans on implementing this? would be very useful. Had plans on using 'UrBackup' with Google Drive mount but as it's built completely on symlinks, it don't seem possible :-(

**Comment by mikwaluk** [comment]
It would be great if you could add this feature! :)

**Comment by dinoboy197** [comment]
> I think this would have to be a new flag for mount, say `--use-links` or something like that which inteprets the `.rclonelink` files and turns them back into symlinks.
> 
> Therefore you would be able to mount your drive remote, and have the symlinks appear as symlinks.
> 
> Is that what is required do you think?

Hi @ncw - from rclone's documentation for `local` mode:

**Comment by dinoboy197** [comment]
I've created a draft PR of a code addition to have the mount command support symbolic links. No extra mount parameters are needed; this works without any extra `--links` command line arguments.

I still need to add documentation and tests, but it should be ready for use. For anyone who is still interested in this feature, please try out [my branch](https://github.com/dinoboy197/rclone/tree/traack_symlinks) and let me know if that works for you! I hope to have the PR ready for review in the next week or two.

**Comment by ncw** [comment]
> I've created a draft PR of a code addition to have the mount command support symbolic links.

Nice one :-)

> No extra mount parameters are needed; this works without any extra `--links` command line arguments.

I'm not sure about that, but we can discuss more!

> I still need to add documentation and tests, but it should be ready for use. For anyone who is still interested in this feature, please try out [my branch](https://github.com/dinoboy197/rclone/tree/traack_symlinks) and let me know if that works for you! I hope to have the PR ready for review in the next week or two.

I note you've done this in the mount layer. There are 3 mount backends - mount/cmount/mount2 and this only applies to one of them. It would be better to do this in the VFS layer which will work with all mount backends. The mount backends will still need a tweak to recognise symlinks though.

**Comment by eatnumber1** [comment]
For folks following along like myself, the PR is #4933. (thanks for working on this!)

**Comment by dinoboy197** [comment]
> There are 3 mount backends - mount/cmount/mount2 and this only applies to one of them.

Wow - I wondered a bit about those, and didn't realize those were any more than prototypes. Thanks for the note!

> It would be better to do this in the VFS layer which will work with all mount backends. The mount backends will still need a tweak to recognise symlinks though.

I doubt that will be any issue, though I admit that I'm not quite sure how the other two mount backends that I didn't tweak actually get enabled. Would you be able to point me to any existing documentation or tips on how to enable the two alternate mount backends so that I can be sure that any changes work with those?

**Comment by ncw** [comment]
> > There are 3 mount backends - mount/cmount/mount2 and this only applies to one of them.
> 
> Wow - I wondered a bit about those, and didn't realize those were any more than prototypes. Thanks for the note!

No worries!

Note also that the VFS layer is used for the rclone servers too, eg `rclone serve http/sftp/webdav`
 
> > It would be better to do this in the VFS layer which will work with all mount backends. The mount backends will still need a tweak to recognise symlinks though.
> 
> I doubt that will be any issue, though I admit that I'm not quite sure how the other two mount backends that I didn't tweak actually get enabled. Would you be able to point me to any existing documentation or tips on how to enable the two alternate mount backends so that I can be sure that any changes work with those?

Sure...

You use `rclone mount ...`, `rclone mount2 ...` and `rclone cmount ...`.

You need to supply the build tag `-tags cmount` to build with cmount support, and you'll need the FUSE headers and a C compiler. cmount uses cgo with the FUSE library, whereas mount talks straight to the kernel.

Note that `rclone mount` is the default mount for linux/bsd/other whereas `rclone cmount` is the default mount for Windows/macOS as `rclone mount` doesn't work there.

At some point I plan to drop mount2 but rclone needs mount and cmount!

**Comment by dinoboy197** [comment]
After some initial refactoring, I realized that symlink support down into the VFS layer means building symlink support for several other components - not just `mount` and `cmount`. I realized that I would need to make some fairly significant design choices regarding how to represent symlinks at that layer - should they be a `File`, but have a new `isSymlink` method, or should there be a new type that derives from `Node` entirely; should symlinks support propagate further down chain into the FS layer, etc. While I was aware earlier that my symlink additions for the `mount` command was fairly trivial and somewhat superficial, I didn't have a proper appreciation for what all would go into proper support in the VFS layer. Not surprising considering I'm a first-time `rclone` contributor!

I don't believe I have the right experience or vision to be making these kinds of changes in the VFS layer as a first-time contributor to `rclone` - I'd have to ask many questions and take up a considerable amount of time trying to sync with the future vision of VFS in order to come up with a solution that I thought was satisfactory. Rather, I'd feel much more comfortable if someone with more of an ownership stake in the VFS architecture and vision implemented the symlink support they'd like to see in that layer, then I'd be comfortable retrofitting my `mount` solution to use whatever VFS hooks were built out.

I say this not a way to suggest that my changes should go to the `rclone` mainline just because I'd rather not implement things in the VFS layer. Rather, I do think that VFS support is the right way to implement this and look forward to someone with more experience (both with `rclone` and with Golang) to drive that a bit. I'm sorry if this comes as a disappointment for anyone waiting for a full solution.

I'll keep an eye out here, and should I become more well-versed with the codebase and VFS direction at a later date, perhaps will pick this up again!

**Comment by ivandeex** [comment]
The attempt PR #4933 was abandoned and closed.
New developers can pick it by the link above and use its patch as a basis.

**Comment by pasnox** [comment]
> > I've created a draft PR of a code addition to have the mount command support symbolic links.
> 
> Nice one :-)
> 
> > No extra mount parameters are needed; this works without any extra `--links` command line arguments.
> 
> I'm not sure about that, but we can discuss more!
> 
> > I still need to add documentation and tests, but it should be ready for use. For anyone who is still interested in this feature, please try out [my branch](https://github.com/dinoboy197/rclone/tree/traack_symlinks) and let me know if that works for you! I hope to have the PR ready for review in the next week or two.
> 
> I note you've done this in the mount layer. There are 3 mount backends - mount/cmount/mount2 and this only applies to one of them. It would be better to do this in the VFS layer which will work with all mount backends. The mount backends will still need a tweak to recognise symlinks though.

Hey, I'm trying to *try* to fix this feature request as it is a very needed feature for me.
Unfortunately, I comes from C/C++, and Go is just shadow for me, so i will have to minimal learn it as I need.
Would you wanna give some inputs on what/where would be to be patched in the vfs ?

Thanks,

**Comment by danielhatton** [comment]
On Sun, 13 Feb 2022, Filipe Azevedo wrote:

> Hey, I'm trying to *try* to fix this feature request as it is a very
> needed feature for me.  Unfortunately, I comes from C/C++, and Go is
> just shadow for me, so i will have to minimal learn it as I need.
> Would you wanna give some inputs on what/where would be to be
> patched in the vfs ?

In case the question was addressed to me: I'm not at all familiar with
the rclone source code (and indeed I no longer have access to a cloud
storage service on which to test anything).  Sorry.

**Comment by pasnox** [comment]
it was for @ncw

**Comment by pasnox** [comment]
@dinoboy197 What's the status of your fork in regards to mount supporting symlink ?

**Comment by dinoboy197** [comment]
Last time I rebased my `dinoboy197/traack_symlinks` branch from rclone master, it brought in commits from 2022-02-18.

**Comment by ncw** [comment]
@dinoboy197 do you want to have another go at sending a PR for this? I like the concept. Most of the code you've done in `cmd/mount` needs to go into the `vfs` - then it will be usable for all users of the vfs, most importantly `cmd/cmount` which is how mount works on macOS and Windows.

**Comment by dinoboy197** [comment]
I appreciate the encouragement! I don't have the time to dedicate to the refactor right now, but it is good to know what the appropriate direction would be if I take this up in the future.

**Comment by rkzed94** [comment]
really looking forward for this to be implemented..
the only missing feature that i need is symlink support for rclone mount.

**Comment by Bra1nsen** [comment]
sorry I don't see through here completely. is it now possible to set hardlinks via rclone and gdrive? and if so how?

**Comment by ncw** [comment]
Hardlinks aren't possible through rclone at the moment.

**Comment by nyuszika7h** [comment]
I would love to see this for both `mount` and `serve` as well, so yeah, it would be best if it was implemented on the VFS layer.

**Comment by pasnox** [comment]
@ncw I'm trying to implement changes for this feature, but I will need some inputs.
Right now i'm struggling to find which part of the code received the remote data that is then used to fill the vfs, so that I can hook the part to add symlinks support. Any idea where I would dig in ?

**Comment by pasnox** [comment]
I got the first bits working for vfs / mount / symlinks - can mount and see / read / open symlinks files on linux.
I will continue the work when back from vacations - around July 22th.

**Comment by ofifoto** [comment]
ooh exciting! some of the git repos I stored in a mount on Google Drive (local space constraints) have symlinks and git wasn't super duper happy about them being missing thanks @pasnox !

**Comment by hongkongkiwi** [comment]
This could be super useful for torrents right? Download to the local drive, move to the VFS mount then symlink back for seeding

**Comment by pasnox** [comment]
I pushed a first test PR @ https://github.com/rclone/rclone/pull/6337
Please, help me debug / test other platforms and backends.

**Comment by pasnox** [comment]
You can grab test binaries at this address: https://beta.rclone.org/branch/pasnox-symlink/v1.60.0-beta.6493.aab476684.pasnox-symlink/

**Comment by biji** [comment]
looks good, mounted using `rclone mount -l`

![image](https://user-images.githubusercontent.com/665559/198569739-07c0c4bb-55a8-4f32-8180-c1d3197b62ae.png)

**Comment by nimagp** [comment]
Hi
Sorry for jumping in, but are binaries usable? 
Will changes be merged into mainline rclone?

**Comment by pasnox** [comment]
> Hi Sorry for jumping in, but are binaries usable? Will changes be merged into mainline rclone?

Latest builds can be found here https://beta.rclone.org/branch/pasnox-symlink/
And yes, of course, goal is to mainstream the changes.

If somebody could test it on windows that would be helpful by the way :)
I can only test macOS / Linux

**Comment by nimagp** [comment]
Thanks.
I'll try and see if I can :)

**Comment by nimagp** [comment]
Can you please take a look at this? @pasnox 
![IMG_20230105_231135](https://user-images.githubusercontent.com/71447191/210866551-eb12686a-1eea-4521-ad72-80e9446c99e4.jpg)
Note : /nix is a rclone mounted storage (using nix-user-chroot) 
And are there any plans on implementing a system like this for saving file permissions? 
I know it would be more difficult as it needs to be updated with every perms change and on mount init (wouldn't take very much resource to set-up all files?)
Can a work-around be reading each file permission only when it's needed and cache them for a period of time?

**Comment by lahma0** [comment]
> If somebody could test it on windows that would be helpful by the way :) I can only test macOS / Linux

@pasnox First, let me just give a big **"THANK YOU"** to you and the rest of the rclone contributors who worked on this. I've been waiting on this feature for a long time but I didn't want to pester the team with yet another feature request; especially a feature such as this which I'm sure will go unused by the vast majority of users. I really need to just get over my hesitancy to learn yet another programming language and get comfortable with Go so I can contribute to the project myself. Anyways, on to the topic at hand.

I downloaded your custom build "_rclone-v1.62.0-beta.6680.19c6081de.pasnox-symlink-windows-amd64.zip_" and tested it on Windows. Here are my results:

**Working:**
Any relative symlinks pointing to files or directories residing in the same directory as the symlink work fine (i.e. test-symlink.txt --> test.txt).

**Not Working:**
Any relative symlinks pointing to files or directories in a parent directory or a subdirectory do **not** work. It appears this is a result of an invalid directory separator as when I list the directory containing the symlink in Powershell 7 (which identifies symlinks and shows what location they are pointing to), the character where the directory separator should be is represented by an invalid character symbol (\uF05C: Invalid Character). Is it possible that maybe you hardcoded the directory separator to "/" instead of choosing the correct separator based on the platform ("\\" in the case of Windows)? Also, it appears that this issue isn't related just to File Explorer or Powershell. I tried a wide range of GUI/console applications and none of them are able to open these invalid symlinks. I've included an image below to further illustrate the issue. Hopefully it is helpful.
![image](https://user-images.githubusercontent.com/4186018/217441008-8d008475-40ca-46dd-bb97-fcbc22df25a2.png)

Again, thank you so much for working on this. I'm lucky I decided to search for this topic again. Needless to say, I was **<ins>very</ins>** surprised to find that someone was actually working on implementing this. Now that I'm aware of the ongoing work on it, I will certainly pay close attention to this thread and as soon as an update is posted, I will test it as thoroughly as possible on Windows and report back with the results.

**Comment by pasnox** [comment]
@lahma0 Thanks for your feedback, I will try to find some times to fix write / read so that it works correctly on windows.
Creating symlinks should be supported yes, but to be honest I do not have any windows OS at hands, so windows support is low on my mind.
Could you tell me what was the command line you tried to create a symlink ?
At a first glance, to fix the write / read of link content, you would have to  fix the overrides in the 3 mount backends, should not be too hard.

**Comment by lahma0** [comment]
> @lahma0 Could you tell me what was the command line you tried to create a symlink ?
Sure, I'm happy to help out in any way that I can as I would love to see this feature working on both Linux and Windows so that it can be pushed to the mainstream build and supported officially. To create a symlink on Windows, you use the "mklink" command.

Here is an example of creating a symlink to a file named "file.txt" stored in the parent directory (relative to cwd):
mklink file-symlink.txt ..\\file.txt  [Note: You must use "\\". If you use "/", the symlink will be invalid on Windows.]

Here is an example of creating a symlink to a directory named "folder" stored in the parent directory (relative to cwd):
mklink /D folder-symlink ..\\folder [Note: You must use "\\". If you use "/", the symlink will be invalid on Windows.]

I'm sure my super long message from before was kind of hard to follow so let me just reiterate the most important points:

- After running 'rclone sync --link' on any Windows storage containing existing symlinks, if you then run 'rclone mount --link', on the destination storage that rclone sync'd to, all of the symlinks will be broken IF the symlink target path contained a directory separator. 
- The reason the symlinks are broken is because rclone is storing the symlink target path in the ".rclonelink" file using Windows directory separators ("\\") instead of Unix directory separators ("/").
- If you manually alter the paths inside the ".rclonelink" files, swapping all Windows directory separators ("\\") with Unix directory separators ("/") and then remount the storage with "--links", all of the symlinks will then be working perfectly fine. 

I won't go into any more details (for now) about trying to create symlinks on an active rclone/vfs mount volume as I think that will just make things more confusing (as the behavior is pretty strange). That can be tackled after fixing the problems with existing symlinks.

If you need access to a Windows VM for testing, I can set one up and send you the address and credentials for it. Just let me know what remote access protocol you prefer (VNC, TeamViewer, RDP, whatever). Or, if you'd prefer, I can just test things for you. Just let me know. I'm happy to help however.

**Comment by lahma0** [comment]
> 

Can you clarify what you mean by "in and outside" symlinks? Are you referring to symlinks that already exist prior to mounting and symlinks that are created on the actual mounted VFS file system?

**Comment by lahma0** [comment]
> Good to know, does that fix both in and outside symlinks ?

@pasnox, when referring to "in and outside" symlinks you mean relative symlinks that point to a file/directory in the same directory (in) or a different directory (outside), then yes, with that one change in the code, all symlinks work fine on Windows.

**Comment by pasnox** [comment]
Right, that was what I had in mind.

**Comment by lahma0** [comment]
> Right, that was what I had in mind.

@pasnox: In contrast to my hackish quick fix, how would you recommend changing the code in order to achieve the same thing in a more permanent, production-ready environment? It would be really great to see this feature fully implemented such that it can be adopted and integrated into the mainline build once and for all. I'm still thinking it might be better to store the symlink paths (within ".rclonelink" files) with Unix directory separators regardless of platform. I'll do a bit more testing when I have some time.

**Comment by pasnox** [comment]
> @pasnox: In contrast to my hackish quick fix, how would you recommend changing the code in order to achieve the same thing in a more permanent, production-ready environment? 

I pushed a fix that would ensure to write unix based file path for symlinks and also ensure to convert them to unix as well when reading symlinks target.
Please pull the new changes and tell me if it's working as expecting with .rclonelink file both containing windows and unix separators.
If yes, then we can go further and recap what does not works when creating symlinks and fix the issues.

**Comment by lahma0** [comment]
> @lahma0 I pushed a fix that would ensure to write unix based file path for symlinks and also ensure to convert them to unix as well when reading symlinks target. Please pull the new changes and tell me if it's working as expecting with .rclonelink file both containing windows and unix separators. If yes, then we can go further and recap what does not works when creating symlinks and fix the issues.

Thank a lot @pasnox. That is awesome! I will check out the new build and report back with the results asap.

**Comment by lahma0** [comment]
@pasnox: I cloned your repository at "https://github.com/pasnox/rclone.git" and ran the same tests again and here are the results:

- As far as I can tell, existing symlinks that are copied/synced to an rclone storage destination (when using "--links") work perfectly fine now. I tried files, directories, symlinks to files/directories in the same directory, symlinks to files/directories in parent/child directories, etc and so on.
- Correct me if I'm wrong, but I believe your previous post implied that you were intending for your code changes to cause symlinks to be stored (in ".rclonelink" files) with Unix directory separators, regardless of platform, right? If that was indeed what you intended, unfortunately that is not what is actually happening. When mounting an rclone storage on Windows (without "--link"), and inspecting the contents of any ".rclonelink" file in a text editor, those paths are still using Windows directory separators ("\\"). Just to be clear, these are fresh, newly created symlinks that were synced to the rclone storage with this new rclone build. Here is an example path from one of those files:
`subdir\test2.txt`
- In regards to creating new symlinks on an actively mounted rclone storage on Windows (which was mounted with "--links"), only the creation of directory symlinks work. Creating directory symlinks using "mklink /D [symlink] [target]" works perfectly fine, regardless of whether the target is in the same directory or a parent/child directory. File symlinks however, do **not** work ("mklink [symlink] [target]"). The following error is returned whenever attempting to create any file symlink: "Cannot create a file when that file already exists."

Please let me know if there is any more information I can provide or anything else you would like for me to test. Otherwise, I will patiently wait for your next build to test. Thanks @pasnox!

**Comment by lahma0** [comment]
> @pasnox: I cloned your repository at "https://github.com/pasnox/rclone.git" and ran the same tests again and here are the results:
> * [...]
>
> Please let me know if there is any more information I can provide or anything else you would like for me to test. Otherwise, I will patiently wait for your next build to test. Thanks @pasnox!

I'm really busy at the moment but when I get some time, I'm going to jump into the code again and see if I can resolve some of these issues myself. @pasnox has obviously already put a lot of time and effort into getting this feature to work and I suspect it is hard to get motivated to fix problems for a platform you don't even use. Since he appears to be busy at the moment too, hopefully I can fix some of these ongoing issues and learn something in the process (although I will almost certainly need at least some input/guidance from @pasnox at some point). I'm hoping to have some time next week to work on this.

**Comment by pasnox** [comment]
> @pasnox: I cloned your repository at "https://github.com/pasnox/rclone.git" and ran the same tests again and here are the results:

Sorry, I though I already answered this.

> 
> * As far as I can tell, existing symlinks that are copied/synced to an rclone storage destination (when using "--links") work perfectly fine now. I tried files, directories, symlinks to files/directories in the same directory, symlinks to files/directories in parent/child directories, etc and so on.

Those existing links were having mixed / or \ separators or at least windows separators ?

> * Correct me if I'm wrong, but I believe your previous post implied that you were intending for your code changes to cause symlinks to be stored (in ".rclonelink" files) with Unix directory separators, regardless of platform, right? 

Right, you can see the changes here https://github.com/pasnox/rclone/commit/a0e00b128dd74c47b564d6a249e95928e8034060#diff-dc21b04023e47c90ae035d0da151cc1feaa216891a4c96c88af5eaf6702e4e08R743
Both writing / reading symlink target replace windows separators by unix separators.

> If that was indeed what you intended, unfortunately that is not what is actually happening. When mounting an rclone storage on Windows (without "--link"), and inspecting the contents of any ".rclonelink" file in a text editor, those paths are still using Windows directory separators ("\"). Just to be clear, these are fresh, newly created symlinks that were synced to the rclone storage with this new rclone build. Here is an example path from one of those files:
>   `subdir\test2.txt`

Did not you said only links to directories were working ?
To be clear, any newly created links on your native file system and synced to rclone, would not transcode separators.
Only links created from a mounted rclone storage would translate separators.

> * In regards to creating new symlinks on an actively mounted rclone storage on Windows (which was mounted with "--links"), only the creation of directory symlinks work. Creating directory symlinks using "mklink /D [symlink] [target]" works perfectly fine, regardless of whether the target is in the same directory or a parent/child directory. 

Great

> File symlinks however, do **not** work ("mklink [symlink] [target]"). The following error is returned whenever attempting to create any file symlink: "Cannot create a file when that file already exists." 

Are you sure about the error message ? I can't find any occurrence of **Cannot create a file when that file already exists** in the entire rclone code base, so it mean one would have to start rclone in a debugger and check where it break.
@ncw Any idea ? My go knowledge is far from good and i'm using a limited environment (Kate text editor + go lsp).

> Please let me know if there is any more information I can provide or anything else you would like for me to test. Otherwise, I will patiently wait for your next build to test. Thanks @pasnox!

I will :)

**Comment by pasnox** [comment]
> I'm really busy at the moment but when I get some time, I'm going to jump into the code again and see if I can resolve some of these issues myself. @pasnox has obviously already put a lot of time and effort into getting this feature to work and I suspect it is hard to get motivated to fix problems for a platform you don't even use. Since he appears to be busy at the moment too, hopefully I can fix some of these ongoing issues and learn something in the process (although I will almost certainly need at least some input/guidance from @pasnox at some point). I'm hoping to have some time next week to work on this.

I will be happy to help you, still there are things I can't do on myself, like starting rclone in a debugger on windows :(.
When we know what cause the issue, we could fix it quickly, and hopefully we could have all platforms covered by symlinks.

**Comment by lahma0** [comment]
> I will be happy to help you, still there are things I can't do on myself, like starting rclone in a debugger on windows :(. When we know what cause the issue, we could fix it quickly, and hopefully we could have all platforms covered by symlinks.

Hey @pasnox, sorry for the long delay. Had to attend an out of state wedding (was the best man) and since I returned, I've been trying to catch back up with work stuff. Been a bit crazy. Anyways, I apologize for my previous report as I think it was obtuse and not very clear. Let me try to be a bit clearer this time.

**Original Scenario:**
First, **ALL** new/existing symlinks that are synced/copied to an rclone storage volume work perfectly. For example, say I have 5 existing symlinks of differing types (symlinks to files/directories, relative symlinks, absolute symlinks, etc) on my local drive in the folder "C:\\temp" and I copy those to an rclone Google Drive storage volume named "gdrive:" using the following command:
`rclone sync c:\temp gdrive:temp --links`
If I then mount that Google Drive storage volume using the command:
`rclone mount gdrive: r: --links`
All of the symlinks that were copied to "gdrive:temp" will show up in the newly mounted rclone storage volume at "r:\\temp" and work exactly as they should without any problems whatsoever. 

Continuing on from the above scenario, let's split into 2 different paths to demonstrate 2 different scenarios:

**Scenario #1:** 
Now say we kill the mount previously discussed and run the exact same mount command again except without the "--links" parameter:
`rclone mount gdrive: r:`
If we navigate to "r:\\temp" and open the file "r:\\temp\\test2-symlink.txt.rclonelink" in a text editor, the path contained within that file is:
`subdir\test2.txt`
As you can see, despite me syncing that symlink to the "gdrive:" storage with this new build, the path is still being saved to the ".rclonelink" file using a Windows directory separator. Whether or not this is actually a problem is debatable since the symlink is working 100% correctly when the storage is mounted using "--links". Just thought I'd mention it since it appears that the intention was to save these paths using Unix directory separators, regardless of platform. 

**Scenario #2:** 
Continuing on from the original scenario, we still have "gdrive:" mounted to "r:" using the "--links" parameter. First, imagine there is a normal text file located at "r:\\temp\\test.txt". Now, imagine we want to create a **new** symlink to that text file (remember, "r:" is a mounted rclone storage volume, not a local drive) at "r:\\temp\\test-symlink.txt". To do that, we open a command prompt, CD to "r:\\temp", and run the command:
`mklink test-symlink.txt test.txt`
The command fails with the error message "Cannot create a file when that file already exists." Note, I've confirmed this is a Windows error message and that's why you aren't finding that error string anywhere in the rclone code base. If I run this exact same command on my local storage volume at "c:\\temp", it works just fine and creates the expected symlink.

Continuing on in scenario #2: Now imagine there is a directory named "subdir" in "r:\\temp" ("r:\\temp\\subdir"). If I instead run the command:
`mklink /D subdir-symlink subdir`
That command will successfully create a valid directory symlink named "subdir-symlink" without any problems. 

**Conclusion:**
So, in other words, creating new directory symlinks on a mounted rclone storage volume works perfectly fine but creating new **file** symlinks does **not**. Additionally, syncing/copying existing symlinks (both file and directory symlinks) on a local storage volume to an rclone storage volume works perfectly (despite the paths being stored in ".rclonelink" files using Windows directory separators). Hopefully this illustrated what I was attempting to say previously in a much more concise manner. Basically, the only thing that really needs to be fixed at this point is creating **file** symlinks on an active rclone storage mount. I suspect the issue is likely related to symlinks needing to be declared explicitly as **file** or **directory** type symlinks on Windows as opposed to having no such distinction on Unix based platforms. Perhaps such a distinction needs to somehow be made in "cmd -> cmount -> fs.go -> Symlink"? I really don't know since I have no experience with FUSE. Please let me know if I can be of any further help!

**Comment by lemaitre-aneo** [comment]
Hi all, I am really interested in this feature and would like to know if you have an estimation about when it would be ready. Would it be a few month, a year, more ?
I'm not in a rush, I just would like to know if I can plan to use it, or if I have to implement a workaround.

Thanks!

**Comment by pasnox** [comment]
> **Conclusion:**
> So, in other words, creating new directory symlinks on a mounted rclone storage volume works perfectly fine but creating new **file** symlinks does **not**. Additionally, syncing/copying existing symlinks (both file and directory symlinks) on a local storage volume to an rclone storage volume works perfectly (despite the paths being stored in ".rclonelink" files using Windows directory separators). Hopefully this illustrated what I was attempting to say previously in a much more concise manner. Basically, the only thing that really needs to be fixed at this point is creating **file** symlinks on an active rclone storage mount. I suspect the issue is likely related to symlinks needing to be declared explicitly as **file** or **directory** type symlinks on Windows as opposed to having no such distinction on Unix based platforms. Perhaps such a distinction needs to somehow be made in "cmd -> cmount -> fs.go -> Symlink"? I really don't know since I have no experience with FUSE. Please let me know if I can be of any further help!

Sorry for the long delay.
I will try to find some time to setup a windows VM and test it on my own as well, it looks strange file only got issues but it's possible some minor details needs to be filled somewhere to properly support it.

**Comment by lemaitre-aneo** [comment]
> > Hi all, I am really interested in this feature and would like to know if you have an estimation about when it would be ready. Would it be a few month, a year, more ? I'm not in a rush, I just would like to know if I can plan to use it, or if I have to implement a workaround.
> > Thanks!
> 
> It depends much how @ncw see the windows support critical or not. If not I think it could land in not so much of time, while windows support could be polished at a second hand.

Thanks for your reply. Personally, I am interested in the Linux support only, so I will be super happy for this to be split in two PRs to make Linux support integrated more quickly. I would understand if this would be kept as a single PR for parity nonetheless.

**Comment by lahma0** [comment]
Really, for all intents and purposes, I think the feature is working well enough on Windows to be merged into the mainline branch. The only part that is actually broken is making **file** symlinks on an actively mounted rclone volume. All symlinks present on local storage prior to syncing/copying work perfectly and I suspect this is all that 95% of people care about. While making file symlinks on a mounted rclone volume is certainly useful, I really don't think it is critical. Perhaps the feature could be integrated as it is now, triggering a warning log message on Windows when someone tries to make a file symlink and then later on, when @pasnox has some time to look into the problem with a Windows VM, that issue could be fixed and the log message removed. Just an idea... I'm really anxious to see this merged into the mainline branch as well!

**Comment by sharkopen** [comment]
@pasnox The 'Cannot create a file when that file already exists' error occurs marked in red when Rename *.rclonelink to the correct filename in vfs/dir.go, 
![1684404245741](https://github.com/rclone/rclone/assets/22067179/6aeb736e-8de0-43b0-aa47-b42c9c67b3a3)

**Comment by sharkopen** [comment]
@pasnox When executing the 'mklink [symlink] [target]' command, the current function call sequence in cmount/fs.go is createEx ---> Getattr ---> Symlink ---> Rename ---> Unlink; the correct sequence should be createEx ---> Getattr ---> Symlink ---> Unlink ---> Rename

**Comment by pasnox** [comment]
@sharkopen Thank you for the extra debug information. I'm very busy those days but I ill try to find some time to at least rebase the git branch, and have a look at the code path call sequence.
Mainwhile, if any (@lahma0 , @sharkopen ) is able to provide a fix for it I will be happy to integrate it fastly.

**Comment by pasnox** [comment]
@sharkopen just curious, why would symlink requires renaming code path ?

**Comment by sharkopen** [comment]
@pasnox You can take a look at the source code of fsp_fuse_intf_SetReparsePoint function in the WinFsp project: https://github.com/winfsp/winfsp/blob/master/src/dll/fuse/fuse_intf.c. In fact, it first creates a hidden file and writes the path of the file or folder that needs to be symlinked into this hidden file. Then it renames the correct path. After that, reading the content of this file will give you the correct path to access. As for why rename is used, there is an explanation in the comments of the source code. 
![image](https://github.com/rclone/rclone/assets/22067179/1c7912ba-8110-4ada-bd50-897069b56e07)

**Comment by lahma0** [comment]
> @sharkopen Thank you for the extra debug information. I'm very busy those days but I ill try to find some time to at least rebase the git branch, and have a look at the code path call sequence. Mainwhile, if any (@lahma0 , @sharkopen ) is able to provide a fix for it I will be happy to integrate it fastly.

Ah, very good! Thanks for pointing out where the exception is occurring @sharkopen. I should be able to implement a fix for it now on behalf of @pasnox. I'll try to take care of it tomorrow and return here with the results. Hopefully we can get this last bug smashed and finally get this integrated into mainline.

**Comment by lahma0** [comment]
So I tried to figure out how to implement a fix for this using the information provided by @sharkopen but I was ultimately unsuccessful. The information provided by @sharkopen however appears to be dead on. The order of operations is indeed createEx ---> Getattr ---> Symlink ---> Rename ---> Unlink. Despite tracing the order of operations in the debugger and monitoring the call stack though, I was never able to actually figure out what parent function was actually responsible for invoking each of those operations (createEx, Getattr, Symlink, etc) in the order they are being called. Perhaps it is WinFSP which is actually invoking these operations? Obviously, I do not have a very good understanding of FUSE/WinFSP/rclone and how they all interact with each other so I'm trying to tackle this problem without much of the understanding that would very helpful. Perhaps @sharkopen or @pasnox might be able to point me in the right direction? If not, I'll mess with it some more when I have time and see if I can't figure it out myself.

**Comment by pasnox** [comment]
I setup a Widows 10 VM with VSCodium, Go, Git. Everything works fine, still i'm unable to run mount commands.
Any idea what i'm missing ?

```
Starting: C:\Users\Filipe Azevedo\go\bin\dlv.exe dap --listen=127.0.0.1:50381 from C:\Projects\Personal\rclone.git
DAP server listening at: 127.0.0.1:50381
Type 'dlv help' for list of commands.
Error: unknown command "mount" for "rclone"
Run 'rclone --help' for usage.
You could use 'rclone selfupdate' to get latest features.

2023/06/10 11:04:24 Fatal error: unknown command "mount" for "rclone"
Process 1312 has exited with status 1
Detaching
```

**Comment by pasnox** [comment]
@lahma0 Hey, I was able to build and cmount a test remote. I can see files and symlinks. Unfortunately for me, clicking any symlinks does not works.
Could you tell me how you made your setup ?

**Comment by lahma0** [comment]
Sorry @pasnox, I've been out of town and just saw your msgs. Glad to see you trying to tackle the Windows issue yourself! I know for Linux users, that is probably about as pleasant as the sound of fingernails on a chalkboard. Looks like you got everything figured out before I was able to reply but I just wanted to check and see if there was anything you were still having issues with. If there is anything I can help with, please let me know. Oh, and just to remind you: despite the intention of saving all paths in .rclonelink files with Unix directory separators, for whatever reason, they are still being saved with Windows directory separators (when created on Windows). Not sure that really matters but I guess maybe it could if switching between Windows and Linux on the same mount? Not really sure..

**Comment by pasnox** [comment]
> @pasnox When executing the 'mklink [symlink] [target]' command, the current function call sequence in cmount/fs.go is createEx ---> Getattr ---> Symlink ---> Rename ---> Unlink; the correct sequence should be createEx ---> Getattr ---> Symlink ---> Unlink ---> Rename

@sharkopen The rclone call sequence seems to match winfsp ?

ends with rename --> unlink as seen here: https://github.com/winfsp/winfsp/blob/master/src/dll/fuse/fuse_intf.c#L2452
Folders are removed before renaming though.
Am I missing something ?

**Comment by pasnox** [comment]
@lahma0 @sharkopen Finally pushed fixes for the windows platform. So far works fine for me. Please test and give feedback, thanks.

**Comment by lahma0** [comment]
That's great news @pasnox! I will rebuild rclone with your most recent changes, test everything out, and report back here later with the results.

**Comment by lahma0** [comment]
Hey @pasnox, have you tried remounting both with/without the --links parameter to check if things are still working after doing so? Because I'm getting some really weird results now... Symlinks created on a mount (that was mounted with --links) are not showing up as .rclonelink files when mounted without --links. Furthermore, whenever I remount with --links, the symlinks I previously created are no longer there. 

Despite that, now all types of symlinks appear to work when created on an active mount but, like I said, they only work during that active mount session.. They are no longer there when remounted.

When remounting, my log is showing tons of entries for any existing symlinks.. They look like this:

`2023/06/12 05:08:50 ERROR : temp3/subfolder02.rclonelink: Failed to read link size: readlink \\?\N:\.cache\rclone\vfsMeta\td-cc-crypt\temp3\subfolder02: The system cannot find the file specified.`

**Comment by pasnox** [comment]
@lahma0 I tried both with and without --links. Still I used the vfs cache mode *write* it seems needed to truncate files on windows, not sure why.
I will have another looks with the default cache mode when I get time.

**Comment by lahma0** [comment]
Hmm.. Ya, I'm using cache mode full. That's weird though, because before your most recent patch, symlinks were working just fine using that cache mode (well.. directory symlinks were working fine.. obviously file symlinks were problematic.. (the whole reason needed for the patch in the 1st place)). So clearly the patch changed something which caused this change.. Ugh, the hole just gets deeper. Windows is such a disastrous mess... What's new, right?

**Comment by pasnox** [comment]
Did you tried cache mode write instead of full ?

**Comment by lahma0** [comment]
@pasnox So I tested with an unencrypted remote and cache modes 'full' and 'writes' and the behavior is identical to before. When mounted with '--links', symlinks appear to work perfectly. However, when remounting, either with or without '--links', the symlinks previously created are not there. It's really weird you aren't seeing the same thing on your end... What else can I do to help out?

Here is an example of the type of log output I'm seeing. Very repetitive entries doing the same thing over and over again. This log is from a non-encrypted mount with cache mode 'writes'. I created a file symlink (test_symlink.txt) and a directory symlink (subsubdir-symlink). Like stated before, when remounted, both symlinks are gone.
https://pastebin.com/KRQuREFx

**Comment by pasnox** [comment]
I'm performing new tests right now, I will disclose some pastbin in the end.
As information, here is what I do use and how:

I'm using **winfsp-2.0.23075.msi**
And use those vscode launch commands:

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "cmount --links",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${workspaceFolder}",
            "args": [
                "cmount",
                "storj-s3-pasnox-tests:tests",
                "C:\\Projects\\Personal\\rclone-tests",
                "--verbose=5",
                "--vfs-cache-mode=writes",
                "--links"
            ],
            "buildFlags": "-tags cmount"
        },
        {
            "name": "cmount",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${workspaceFolder}",
            "args": [
                "cmount",
                "storj-s3-pasnox-tests:tests",
                "C:\\Projects\\Personal\\rclone-tests",
                "--verbose=5",
                "--vfs-cache-mode=writes"
            ],
            "buildFlags": "-tags cmount"
        }
    ]
}
```

**Comment by pasnox** [comment]
@lahma0 I pushed a new way to handle winfsp hacky symlinks support, hacky implementation needs... hacky workaround :|
This now support creating symlinks with all cache mode, so please try both off/full.

**Comment by lahma0** [comment]
@pasnox Good deal! I will test it right now. Additionally, I would be happy to hop onto the #rclone IRC channel. Efnet?

**Comment by pasnox** [comment]
> Efnet
libera.net

**Comment by lahma0** [comment]
@pasnox and I have been talking on IRC and doing some testing. As things stand now, the only issue that remains is the creation of new**_file_** symlinks on Windows mounts whenever cache mode is set to 'full'. When cache mode is set to 'off', creation of new file and directory symlinks work fine as well as pre-existing file/directory symlinks. When cache mode is set to 'full', all preexisting symlinks work fine but creation of new file symlinks causes rclone to lock up (creation of new directory symlinks works fine). Except for this one last issue, this new feature appears to be ready to push out to the mainline branch. @pasnox said this last remaining issue is due to a workaround/hack that he cannot fix himself. I think the hack/workaround he is referencing is referenced here in the WinFSP source code:
https://github.com/winfsp/winfsp/blob/6fb72555d33d228736b8156cb919edeaae4da150/src/dll/fuse/fuse_intf.c#L2293|

I think the workaround uses this function to temporarily rename a symlink file which is causing issues when using a cache mode other than 'off':
https://github.com/winfsp/winfsp/blob/6fb72555d33d228736b8156cb919edeaae4da150/src/dll/fuse/fuse_intf.c#L224

My understanding of this could be wrong (if so, perhaps @pasnox could chime in) but ultimately, I think the only way we are going to be able to finally fix this issue and move the new feature into the mainline branch is to get some help from @ncw . @pasnox has clearly stated to me on IRC that he does not know how to fix this issue and that it is going to require help from @ncw in order to fix. Hopefully some further workaround can be implemented to make things work without having to implement changes in WinFSP because I suspect that will only cause things to stretch out further and possibly might mean the fix will never be implemented. I know @ncw is surely a very busy person but hopefully at some point, when he has time, he might be able to offer up his help or advice on fixing this problem. I'd hate to see all this hard work that @pasnox has put into this feature go to waste.

**Comment by lahma0** [comment]
Hey @pasnox, I just wanted to follow up on this and ask if you know whether or not the Windows implementation for this was ever fully fixed and implemented? I see that the documentation on rclone.org now includes the "--links" flag so it appears that the feature has been pushed out to the mainline branch. However, I'm not sure if that is Linux-only support or if that includes Windows too. Once again, thank you so much for all the hard work you put into implementing this. It's so awesome to actually have this in the public build of rclone now (even if it is only for Linux... though hopefully not)!

**Comment by pasnox** [comment]
Hey @lahma0 , no sorry, there was no advance in this area, and @ncw seems busy or just don't want to pass some time on helping fixing it.
On my side i'm really busy and I do not have time to maintains it anymore, and I have to admit that my motivation is now rather low arguing for how long this PR has going on.
Let hope @ncw will help revive it, else I will just let it die :(

**Comment by lahma0** [comment]
Hey @pasnox, thanks for responding to my query. I totally understand your sentiments. You've absolutely gone above and beyond in writing the code necessary to get this feature implemented. I doubt the feature would have been added for a very long time without your time and dedication in seeing it through. Quite frankly, I think you spent an unreasonable amount of time trying to implement the wonky Windows support, especially when you don't use, or intend to use, Windows anyways. I think it is safe to say that me and the rest of the rclone community really appreciate the time and effort spent by you and other contributors to make rclone what it is today. 

While it is sad to see the Windows support fall through the cracks when it is like 95% finished, I guess we will just have to hold out hope that @ncw will eventually have the time and motivation to help out with this. I'm sure he has **plenty** of other more pressing matters to attend to but if he does eventually decide to take a look, at least this thread has most of the relevant information documented. I'm pretty sure I noted, in detail, the results of my testing and exactly what isn't working correctly. For the record, I believe the **only** thing not working in Windows is **file** symlink creation whenever cache mode is enabled (creation of directory symlinks works fine). Literally, that's it.

Do you know if rclone allows people to post bounties/rewards for feature requests or bug fixes? I know I've seen other GitHub projects that do. Just asking because I know I would be willing to chip in some cash to see this finished out. I guess I can just donate to the project and include a note referencing this issue if not. If you don't know, perhaps someone else that reads this and does know might let me know. Anyways, thanks again @pasnox!

**Comment by ncw** [comment]
This feature is approaching ready - please try this beta if possible

[v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks](https://beta.rclone.org/branch/fix-2975-mount-symlinks/v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks/) on branch [fix-2975-mount-symlinks](https://github.com/rclone/rclone/tree/fix-2975-mount-symlinks) (uploaded in 15-30 mins)

Use the `--links` flag on `rclone mount` to enable.

**Comment by pjhfggij** [comment]
fantastic, I'm looking forward for be able to setup python virtual environments directly on rclone mounts

**Comment by hongkongkiwi** [comment]
> This feature is approaching ready - please try this beta if possible
> 
> [v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks](https://beta.rclone.org/branch/fix-2975-mount-symlinks/v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks/) on branch [fix-2975-mount-symlinks](https://github.com/rclone/rclone/tree/fix-2975-mount-symlinks) (uploaded in 15-30 mins)
> 
> Use the `--links` flag on `rclone mount` to enable.

is this in the official beta build installed from the website?

**Comment by ncw** [comment]
> is this in the official beta build installed from the website?

Not yet, please try from the link above

**Comment by darthShadow** [comment]
It may be colliding with the `--links` flag of the local FS. Can you prefix the mount command with `RCLONE_LOCAL_LINKS=true` and see if that works?

**Comment by ncw** [comment]
> It may be colliding with the `--links` flag of the local FS. Can you prefix the mount command with `RCLONE_LOCAL_LINKS=true` and see if that works?

Well spotted @darthShadow 

I'm not sure why that is happening - I would have expected the `--links` flag to be picked up by both the VFS and the local backend.

Assuming I can fix it - I think that is the correct think to do - would you agree?

I did consider making a separate flag `--vfs-links` to separate it from the local backend `--links` flag. But then I was thinking that is probably not what users expect.

Maybe I should stop the `--links` flag being a local backend specific flag and make it a global flag which the local backend reads as does the VFS - that is the usage I was thinking of. That is going to be untidy though!

Thoughts?

**Comment by ncw** [comment]
I've made the `--links` flag global so it controls both the VFS and the local backend here. There are also individual flags `--local-links` and `--vfs-links` if you want to be more selective.

If anyone wants to have a try

[v1.69.0-beta.8433.67d905e40.fix-2975-mount-symlinks](https://beta.rclone.org/branch/fix-2975-mount-symlinks/v1.69.0-beta.8433.67d905e40.fix-2975-mount-symlinks/) on branch [fix-2975-mount-symlinks](https://github.com/rclone/rclone/tree/fix-2975-mount-symlinks) (uploaded in 15-30 mins)


@darthShadow wrote
> Yeah, I think using the same flag would be the right approach. I would even go a step further and suggest we remove the `--links` flag altogether from the local backend and use the `rclonelink` behavior as the default, with an override as now using the `copy-links` flag.

So make `--links` the default effectively for the local backend? That would be a backwards incompatible change.

> I am just not sure if anyone has gotten used to the behaviour of seeing the `.rclonelink` file and may break their scripts/apps if they start seeing the symlink instead.

They will only see the symlink on `rclone mount` everything else will remain as before. You won't see symlinks unless you enable it explicitly with `--links` or `--vfs-links`.

**Comment by darthShadow** [comment]
> That would be a backwards incompatible change.

Yeah, but it would bring it inline with the behavior of the other backends. I am fine either way, just wanted to bring it up once to see if anyone else had any comments on it.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.69

Thank you @pasnox 

Any more testing gratefully received :-) 

BTW this was the resolution to the --links flag problem which gives the most flexibility at the cost of a few more flags (!).

    fs: make --links flag global and add new --local-links and --vfs-links flag
    
    Before this change the --links flag when using the VFS override the
    --links flag for the local backend which meant the local backend
    needed explicit config to use links.
    
    This fixes the problem by making the --links flag global and adding a
    new --local-links flag and --vfs-links flags to control the features
    individually if required.

**Comment by i-api** [comment]
Hi @ncw, sadly it seems symlinks don't seem to work on the mounted "crypt" type.  
I am able to create symlinks inside the mount. However, after I unmount, or sometimes even randomly even if i haven't unmounted, the symlinks are being deleted erroneously.  
Are the flags I'm using wrong?  I'm just mounting local to local.

```
rclone --version                                                                                    
rclone v1.69.0
- os/version: darwin 14.2.1 (64 bit)
- os/kernel: 23.2.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.23.4
- go/linking: dynamic
- go/tags: cmount


rclone mount \
        --crypt-remote ~/my_crypt_dir \
        :crypt: ~/folder_123 \
        --crypt-password  $(rclone obscure "12345") \
        --crypt-password2 $(rclone obscure "12345") \
        --links \
        --metadata \
        --vfs-cache-mode=writes \
        --vfs-cache-max-size=500M \
        --vfs-cache-poll-interval=5s \
        --crypt-filename-encoding "base32768" 

bash-3.2$ cd ~/folder_123 
bash-3.2$ ln -s test.txt test_symlink       # creating a symlink file with ln -s
bash-3.2$ readlink test_symlink              # verify the symlink
test.txt
bash-3.2$ cd ~/
bash-3.2$ ls -la folder_123/
total 16
drwxr-xr-x@  1 a  staff     0 Feb  7 15:50 .
drwxr-x---+ 49 a  staff  1568 Feb  7 15:48 ..
-rw-r--r--   1 a  staff    18 Feb  7 15:50 test.txt
lrw-r--r--   1 a  staff     8 Feb  7 15:51 test_symlink -> test.txt
bash-3.2$ cat folder_123/test.txt
this is test file
bash-3.2$ cat folder_123/test_symlink
this is test file
bash-3.2$

# after unmounting and remounting the symlink file is erroneously deleted?

bash-3.2$ ls -la folder_123/
total 8
drwxr-xr-x@  1 a  staff     0 Feb  7 15:52 .
drwxr-x---+ 49 a  staff  1568 Feb  7 15:48 ..
-rw-r--r--   1 a  staff    18 Feb  7 15:50 test.txt
bash-3.2$

```

**Comment by hongkongkiwi** [comment]
You may be better off to open a new issue and link it to this.

**Comment by ncw** [comment]
@i-api that does look like a bug. Can you put it in a new issue please? Thanks

**Comment by despian** [comment]
I'm also using crypt and it seems to be working as expected for me. Though I'm using the `--vfs-links` switch.

**Comment by i-api** [comment]
> [@i-api](https://github.com/i-api) that does look like a bug. Can you put it in a new issue please? Thanks

sure @ncw  https://github.com/rclone/rclone/issues/8367

**Comment by i-api** [comment]
> I'm also using crypt and it seems to be working as expected for me. Though I'm using the `--vfs-links` switch as well as `--vfs-cache-mode full`.

i tried this thank you sir, unfortunately it didn't work as expected for me

**Comment by pjhfggij** [comment]
> released in v1.69

nice
